exchange/auction_test.go

```diff
@@ -124,62 +142,82 @@ func runCacheSpec(t *testing.T, fileDisplayName string, specData *cacheSpec, bid
+				{
+					Min:       0,
+					Max:       5,
+					Increment: 0.05,
+				},
+				{
+					Min:       5,
+					Max:       10,
+					Increment: 0.1,
+				},
+				{
+					Min:       10,
+					Max:       20,
+					Increment: 0.5,
+				},
+			},
+		},
+		includeWinners:    specData.TargetDataIncludeWinners,
+		includeBidderKeys: specData.TargetDataIncludeBidderKeys,
+		includeCacheBids:  specData.TargetDataIncludeCacheBids || bids,
+		includeCacheVast:  specData.TargetDataIncludeCacheVast || vast,
+	}
+
 	testAuction := &auction{
 		winningBids:         winningBids,
 		winningBidsByBidder: winningBidsByBidder,
 		roundedPrices:       roundedPrices,
 	}
-	_ = testAuction.doCache(ctx, cache, bids, vast, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
-	found := 0
+	_ = testAuction.doCache(ctx, cache, targData, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
 
-	for _, cExpected := range specData.ExpectedCacheables {
-		for _, cFound := range cache.items {
-			// make sure Data section matches exactly
-			if !vast {
-				// not testing VAST XML for custom cache key test
-				eq := jsonpatch.Equal(cExpected.Data, cFound.Data)
-				if !eq {
-					continue
-				}
-			}
+	if !vast {
+		if len(specData.ExpectedCacheables) == len(cache.items) {
+			var ht map[string]int = make(map[string]int)
+			var compareString string
 
-			// make sure Key value is as expected
-			keymatch := false
-			if len(cExpected.Key) > 0 {
-				// can only verify prefix; remainder is random
-				if strings.HasPrefix(cFound.Key, cExpected.Key) {
-					keymatch = true
-				}
-			} else {
-				if len(cFound.Key) == 0 {
-					// Key is expected to be empty
-					keymatch = true
-				}
+			for _, cExpected := range specData.ExpectedCacheables {
+				compareString = string(cExpected.Data) + string(cExpected.TTLSeconds)
+				ht[compareString] += 1
 			}
-			if !keymatch {
-				continue
+			for _, cFound := range cache.items {
+				compareString = string(cFound.Data) + string(cFound.TTLSeconds)
+				ht[compareString] -= 1
 			}
-
-			// make sure TTLSeconds section matches exactly
-			if cExpected.TTLSeconds == cFound.TTLSeconds {
-				found++
+			for _, freq := range ht {
+				if freq > 0 {
+					t.Errorf("%s:  Less elements were cached than expected \n", fileDisplayName)
+				} else if freq < 0 {
+					t.Errorf("%s:  More elements were cached than expected \n", fileDisplayName)
+				}
 			}
 		}
 	}
-
-	if found != len(specData.ExpectedCacheables) {
-		fmt.Printf("Expected:\n%v\n\n", specData.ExpectedCacheables)
-		fmt.Printf("Found:\n%v\n\n", cache.items)
-		t.Errorf("%s:  All expected cacheables not found. Expected %d, found %d.", fileDisplayName, len(specData.ExpectedCacheables), found)
+	for _, cExpected := range specData.ExpectedCacheables {
+		for _, cFound := range cache.items {
+			// make sure Key value is as expected
+			_, err := regexp.MatchString(cExpected.Key+`[a-zA-Z0-9_-]*`, cFound.Key)
+			if err != nil {
+				continue
```

Comment:
I am not sure what are we trying to do here. If for an expectedCacheable, you don't find in the actual cache items a cache key with the regexp cExpected.Key+`[a-zA-Z0-9_-]*`, you just continue and move on to the next expectedCacheable item.

If we want to assert that for every expectedCacheable, there exists a an actual cache item with the cache key that matches the cExpected.Key+`[a-zA-Z0-9_-]*` regexp then we should do something like:

```
for _, cExpected := range specData.ExpectedCacheables {
	cacheKeyFound := false
	for _, cFound := range cache.items {
		// make sure Key value is as expected
		_, err := regexp.MatchString(cExpected.Key+`[a-zA-Z0-9_-]*`, cFound.Key)
		if err != nil {
			continue
		}
		cacheKeyFound = true
		break
	}
	if !cacheKeyFound {
		t.Errorf("%s cache key either not found in the cache result or doesn't match the expected key format.", cExpected.Key)
	}
}
```

Reply:
Mansi this function has been modified. Thanks for pointing out this flaw. Indeed the for loop was not exactly asserting anything and after careful thought, it turns out it was not even needed. Please let me know if you agree with the new approach.

Reply:
@guscarreon I see that the cache key format assertion has now been removed and as per our discussion offline, I don't think I am the best person here to judge if those assertions are needed or not. I'd guess they are but I think @hhhjort is a better person to answer that.

Reply:
Mansi your observations here were very important to correct all test cases in this document